internal/limitreader/reader.go

```diff
@@ -0,0 +1,49 @@
+package limitreader
+
+import (
+	"bufio"
+	"errors"
+	"io"
+)
+
+// ErrReadBeyondLimit is the error returned when there's more data on the
+// underlying reader and a read operation is trying to read beyond that.
+var ErrReadBeyondLimit = errors.New("limitreader: read beyond limit")
+
+// New wraps reader with a limit.
+//
+// It's similar to io.LimitReader,
+// but instead of always returning io.EOF when read beyond the limit,
+// it would return a different error (which would cause the read operation to
+// fail) when there's more data on the underlying reader.
+func New(r io.Reader, limit int64) io.Reader {
+	return &reader{
+		reader: bufio.NewReader(r),
+		limit:  limit,
+	}
+}
+
+type reader struct {
+	reader *bufio.Reader
+	limit  int64
+	read   int64
+}
+
+func (r *reader) Read(p []byte) (int, error) {
+	// EOF handling
+	if len(p) > 0 && r.read >= r.limit {
+		_, err := r.reader.Peek(1)
+		if err != nil {
+			// NOTE: This also includes io.EOF error, which is the expected behavior.
+			return 0, err
+		}
+		return 0, ErrReadBeyondLimit
+	}
+
+	if remaining := r.limit - r.read; len(p) > int(remaining) {
+		p = p[:remaining]
+	}
+	read, err := r.reader.Read(p)
+	r.read += int64(read)
+	return read, err
+}
```

Comment:
I don't think this should introduce buffering.  Leave that to the user.  I think that simplifies this to:

```suggestion
func New(r io.Reader, limit int64) io.Reader {
	return &reader{
		reader: r,
		limit:  limit,
	}
}

type reader struct {
	reader io.Reader
	limit  int64
}

func (r *reader) Read(p []byte) (int, error) {
  if r.limit <= 0 {
    return 0, ErrReadLimitExceeded
  }
	if r.limit > len(p) {
	  p = p[:r.limit]
	}
	n, err := r.reader.Read(p)
	r.limit -= n
	return n, err
}
```

Reply:
I would also 🔕 recommend exporting the LimitReader struct and returning that instead of io.Reader, since that would allow for some amount of inlining at call-sites, though it's unlikely to be relevant for us.

Reply:
The suggestion you proposed is essentially the same as `io.LimitReader`, except that it returns a different error than `io.EOF`. The problem with that approach is that it cannot distinguish between the corner case of the underlying reader has exactly limit bytes of data, vs. the underlying reader has more than limit bytes of data. The reason is that in most of the `io.Reader` implementations, when you read the last n bytes of data, it won't return `n, io.EOF`. They would return `n, nil`, and then return `0, io.EOF` on the next `Read` call. (see https://play.golang.org/p/s8GEFnbn-T- for an example).

We can, of course, always add the limit by 1 in `New`, so that it can distinguish between the two cases. But that will result in this implementation actually could return limit+1 data. That's not a big problem for our use cases, and probably fine if we keep this implementation under `internal/` forever, but that is not something I would ever expose as a public API in the future.

The `bufio` approach is just more correct, although the correctness is not super important here. In the worst case scenario, that the underlying reader has a chunk of exactly limit bytes of data, and the next chunk would block for a bit before returning, both approaches (`bufio` or just use `limit+1`) would hit that blocking Read anyways in order to return the error. `bufio` will not block for longer even if the next chunk does not have enough data to fill the buffer size. The only downside of `bufio` approach is that it uses slightly more memory for the buffer, but the default buffer size is only 4096 bytes (https://cs.opensource.google/go/go/+/refs/tags/go1.16.5:src/bufio/bufio.go;drc=refs%2Ftags%2Fgo1.16.5;l=19).

Reply:
Sure, if you want to deal with the boundary condition, it looks like this:

https://play.golang.org/p/GbaDhtZOu_4
```
func (r *reader) Read(p []byte) (int, error) {
	if r.limit <= 0 {
		n, err := r.reader.Read(p[:1])
		if !errors.Is(err, io.EOF) {
			err = ErrReadLimitExceeded
		}
		return n, err
	}
	if r.limit < int64(len(p)) {
		p = p[:r.limit]
	}
	n, err := r.reader.Read(p)
	r.limit -= int64(n)
	return n, err
}
```

That will try to read a test byte on the off-chance that precisely the amount of data is read.

Reply:
That's still an incorrect implementation. (it's correct "enough" for our use case). If someone keeps calling `Read` after reached the limit, those `Read` calls would eventually return `io.EOF` instead of `ErrReadLimitExceeded`. With `bufio` approach after the every `Read` call is guaranteed to return `ErrReadLimitExceeded` after the first one returning that error.

Reply:
https://play.golang.org/p/Urf1103SLo_8 demonstrates the "correctness" of the different approaches.

I use `io.LimitReader` as the baseline, as that's what we are replacing. We really only want to replace the error returned, not the actual content returned (your approach will read one more byte).

Reply:
The reason that I return the extra byte is because it is the more correct thing to do, because the extra byte is read from the underlying reader.  With your limit reader, any data caught in the buffer is completely lost, which is a far larger sin than returning one extra byte.

Observe: https://play.golang.org/p/KPfLanNM7Ul
```
original content: "123456789", limit: 5

  io.LimitReader: read: "12345", err: <nil>
                  then: "6789", err: <nil>
       no buffer: read: "123456", err: read limit exceeded
                  then: "789", err: <nil>
           bufio: read: "12345", err: read limit exceeded
                  then: "", err: <nil>
```

In any case, the fixes for the two issues you state are trivial:
* Lie about the peeked byte, like you do in the buffered version (but honestly, this is _bad_ and you shouldn't do it)
* Save returned errors so repeated calls are noops (this is only needed if your callers are doing weird things)
https://play.golang.org/p/U9G2mQgWimO

Reply:
It is not at all uncommon to depend on this kind of behavior, either.  For example, if we e.g. read the first 1k of a file, do envsubst on it, and then combine with the rest, using the presence of the error to report to the user a warning that their config was too long.  You can do that with the stdlib limit reader (without the warning), or my limit reader, but not yours because yours discards data.  Readers should not discard data in general, it's a loaded footgun.